lab4/PlayField.cpp

```diff
@@ -0,0 +1,191 @@
+#include "PlayField.h"
+#include <cassert>
+#include <vector>
+
+PlayField::CellPos::CellPos(int cellNum) {
+
+    _row = cellNum / fieldSize;
+
+    _column = cellNum % fieldSize;
+
+}
+
+
+
+PlayField::CellPos::CellPos(int row, int column) : _row(row), _column(column) { }
+
+
+
+int PlayField::CellPos::getRow() const {
+
+    return _row;
+
+}
+
+
+
+void PlayField::CellPos::setRow(int value) {
+
+    assert(value >= 0 && value <= fieldSize - 1);
+
+    _row = value;
+
+}
+
+
+
+int PlayField::CellPos::getColumn() const {
+
+    return _column;
+
+}
+
+
+
+void PlayField::CellPos::setColumn(int value) {
+
+    assert(value >= 0 && value <= fieldSize - 1);
+
+    _column = value;
+
+}
+
+
+
+PlayField::CellPos::operator int() const {
+
+    return _column + _row * fieldSize;
+
+}
+
+
+
+PlayField::CellStatus PlayField::operator[](const CellPos &cellPos) const {
+
+    return _field[int(cellPos)];
+
+}
+
+
+
+std::vector<PlayField::CellPos> PlayField::getEmptyCells() const {
+
+    std::vector<CellPos> result;
+
+    for (int i = 0; i < _field.size(); ++i)
+
+        if (_field[i] == CellStatus::csEmpty)
+
+            result.emplace_back(CellPos(i));
+
+    return result;
+
+}
+
+bool PlayField::checkWinStatus(PlayField::CellStatus player) const {
+
+    return hasWin(player,true) || hasWin(player);
+
+}
+
+
+
+PlayField::FieldStatus PlayField::checkFieldStatus() const {
+
+    if (checkWinStatus(CellStatus::csCross))
+
+        return FieldStatus::fsCrossesWin;
+
+    if (checkWinStatus(CellStatus::csNought))
+
+        return FieldStatus::fsNoughtsWin;
+
+    int crosses = countCells(CellStatus::csCross);
+
+    int noughts = countCells(CellStatus::csNought);
+
+    if (!(crosses == noughts || crosses == noughts + 1))
+
+        return FieldStatus::fsInvalid;
+
+    if (getEmptyCells().empty())
+
+        return FieldStatus::fsDraw;
+
+    return FieldStatus::fsNormal;
+
+}
+
+
+
+PlayField PlayField::operator+(const CellPos &cellPos) const {
+
+    PlayField newField;
+
+    newField._field = _field;
```

Comment:
А почему сразу не присвоить?
```suggestion
    PlayField newField = *this;
```

Reply:
Согласен, переделаю